python/myelin/gradient.py

```diff
@@ -0,0 +1,262 @@
+# Copyright 2017 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+"""Myelin function gradient builder."""
+
+import math
+from sling.myelin.builder import Builder
+
+# z = x + y
+# dx = dz
+# dy = dz
+def add_grad(op, g):
+  x = op.inputs[0]
+  y = op.inputs[1]
+  z = op.outputs[0]
+  g.add(x, g.d(z))
+  g.add(y, g.d(z))
+
+# z = x - y
+# dx = dz
+# dy = -dz
+def sub_grad(op, g):
+  x = op.inputs[0]
+  y = op.inputs[1]
+  z = op.outputs[0]
+  g.add(x, g.d(z))
+  g.add(y, g.expr.neg(g.d(z)))
+
+# z = x * y
+# dx = dz * y
+# dy = x * dz
+def mul_grad(op, g):
+  x = op.inputs[0]
+  y = op.inputs[1]
+  z = op.outputs[0]
+  g.add(x, g.expr.mul(g.d(z), g.v(y)))
+  g.add(y, g.expr.mul(g.v(x), g.d(z)))
+
+# z = x * y
+# dx = dz * y^T
+# dy = x^T * dz
+def matmul_grad(op, g):
+  x = op.inputs[0]
+  y = op.inputs[1]
+  z = op.outputs[0]
+
+  if op.attr("transpose_b"):
+    g.add(x, g.expr.matmul(g.d(z), g.v(y)))
+  else:
+    g.add(x, g.expr.matmul(g.d(z), g.expr.t(g.v(y))))
+
+  if op.attr("transpose_a"):
+    g.add(y, g.expr.matmul(g.v(x), g.d(z)))
+  else:
+    g.add(y, g.expr.matmul(g.expr.t(g.v(x)), g.d(z)))
+
+# z = x / y
+# dx = y * dz
+# dy = (-x / y^2) * dz
+def div_grad(op, g):
+  x = op.inputs[0]
+  y = op.inputs[1]
+  z = op.outputs[0]
+  dx = g.expr.div(g.d(z), g.v(y))
+  g.add(x, dx)
+  dy = g.expr.mul(g.d(z), g.expr.div(g.expr.neg(g.v(x)),g.expr.square(g.v(y))))
+  g.add(y, )
+
+# y = x * x
+# dx = 2 * x * dy
+def square_grad(op, g):
+  x = op.inputs[0]
+  y = op.outputs[0]
+  g.add(x, g.expr.mul(g.d(y), g.expr.mul(2.0, g.v(x))))
+
+# y = 1 / x
+# dx = -1 / x^2 * dy
+def reciprocal_grad(op, g):
+  x = op.inputs[0]
+  y = op.outputs[0]
+  g.add(x, g.expr.mul(g.d(y), g.expr.neg(g.expr.rcp(g.expr.square(g.v(x))))))
```

Comment:
Oops. This is my autograd experiment in python which was included in this PR by mistake.

However, the optimization is still good since you save a division, so I have changed this in the C++ version. Thanks for spotting this.